store/src/server/api.rs

```diff
@@ -134,6 +145,52 @@ impl api_server::Api for StoreApi {
             nullifiers: state_sync.nullifiers,
         }))
     }
+
+    async fn block_inputs(
```

Comment:
> I don't think this should cause any immediate issues, but I wonder if a better approach would be as follows:

Actually there is a race in the code:

- `get_block_header`, returns latest block for height `n`
- concurrent `apply_block` moves the block to `n+1`
- `account_tree` is read, now the state corresponds to `n+1`

The above means all the proofs from `account_tree` are incorrect.

---

With the above said, you and I have a fundamentally different perspective on how to approach this. My recommendation is to have structures that can read old versions of the data, this is currently unsupported and doesn't seem to be the way we are moving forward. If we are going to rely on locking, it probably makes sense to have a global `RWLock` that covers the data structures and the db.

Reply:
My thinking here is that adding structures which keep old data will add significant overhead (both from engineering and performance standpoints) while managing locks shouldn't be too bad for our specific use case because:

1. We don't have concurrent writes (i.e., all writes should be serialized).
2. Writes happen rather infrequently (e.g., once every 3 seconds).
3. All writes go through a single endpoint (`apply_block`) and should be well-encapsulated.

The specific issue with the race condition we can probably resolve when we work on `apply_block` endpoint - so, I'll mark this as resolved for now.

Reply:
I'm going to start the `apply_block` soon, what is the locking discipline you want?